src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,564 @@
+  } else {
+    _apply_to_inputs(node);
+    return;
+  }
+
+  auto right_tree_root = subquery_expression->lqp;
+
+  // We cannot support anti joins right now (see large comment below on multi-predicate joins for the reasons).
+  if (join_mode != JoinMode::Semi) {
+    _apply_to_inputs(node);
+    return;
+  }
+
+  // Do not reformulate if expected output is small.
+  if (node->get_statistics()->row_count() < 100.0f) {
+    _apply_to_inputs(node);
+    return;
+  }
+  std::cout << "node cost before in reformulation: " << CostModelLogical().estimate_plan_cost(node) << '\n';
+
+  // TODO(anybody): Is this check actually necessary, or is this always true for correlated parameters?
+  // Check that all correlated parameters expressions are column expressions of the left sub-tree. Otherwise, we won't
+  // be able to use them in join predicates. Also build up a map from parameter ids to their respective expressions.
+  std::map<ParameterID, std::shared_ptr<AbstractExpression>> parameter_mapping;
+  for (size_t parameter_idx = 0; parameter_idx < subquery_expression->parameter_count(); ++parameter_idx) {
+    const auto& parameter_expression = subquery_expression->parameter_expression(parameter_idx);
+    if (!left_tree_root->find_column_id(*parameter_expression)) {
+      _apply_to_inputs(node);
+      return;
+    }
+
+    parameter_mapping.emplace(subquery_expression->parameter_ids[parameter_idx], parameter_expression);
+  }
+
+  // Find usages of correlated parameters and try to turn them into join predicates. To pull up the predicates safely,
+  // other nodes might need to be adjusted:
+  //    - Projection nodes might remove the required columns, so they are simply removed. Since we join with a
+  //      semi/anti join, this does not change the semantics. We run before ColumnPruningRule, which then re-adds
+  //      appropriate pruning projections.
+  //    - Alias nodes are due to the same reasons as projection nodes. Since we can't remove them (that would remove
+  //      the aliased columns) we simply extend them to only add and not remove any columns.
+  //    - Aggregate nodes also remove columns. For now, we only support pulling equals predicates above them. For that
+  //      to work, we extend the aggregate nodes to also group by the columns referenced by all predicates pulled from
+  //      below to above them.
+  //
+  // The only other nodes we pull predicates past are sort and validate nodes, which don't need to be adjusted.
+  auto pull_up_info = prepare_predicate_pull_up(right_tree_root, parameter_mapping);
+  auto is_correlated_parameter = [&](ParameterID id) { return parameter_mapping.find(id) != parameter_mapping.end(); };
+  if (contains_unoptimizable_correlated_parameter_usages(right_tree_root, is_correlated_parameter,
+                                                         pull_up_info.predicate_nodes)) {
```

Comment:
I changed it to pass `parameter_mapping` around directly (we introduced the lambda at some point because the check was used in a lot of places, now its only used once)